File: scripts/sepolia_rpc_relay.py
#!/usr/bin/env python3
import argparse
import json
import sys
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from urllib.error import HTTPError, URLError
from urllib.request import Request, urlopen


class RelayHandler(BaseHTTPRequestHandler):
    upstream = ""
    timeout = 30

    def do_POST(self):
        length = int(self.headers.get("Content-Length", "0"))
        body = self.rfile.read(length)

        request = Request(
            self.upstream,
            data=body,
            headers={"Content-Type": "application/json"},
            method="POST",
        )

        try:
            with urlopen(request, timeout=self.timeout) as response:
                data = response.read()
                self.send_response(response.status)
                self.send_header("Content-Type", response.headers.get("Content-Type", "application/json"))
                self.send_header("Content-Length", str(len(data)))
                self.end_headers()
                self.wfile.write(data)
        except HTTPError as exc:
            data = exc.read()
            self.send_response(exc.code)
            self.send_header("Content-Type", exc.headers.get("Content-Type", "application/json"))
            self.send_header("Content-Length", str(len(data)))
            self.end_headers()
            self.wfile.write(data)
        except URLError as exc:
            self._write_error(502, f"upstream unavailable: {exc}")
        except Exception as exc:  # pragma: no cover - defensive path
            self._write_error(500, f"relay failure: {exc}")

    def do_GET(self):
        if self.path == "/healthz":
            payload = json.dumps({"ok": True, "upstream": self.upstream}).encode()
            self.send_response(200)
            self.send_header("Content-Type", "application/json")
            self.send_header("Content-Length", str(len(payload)))
            self.end_headers()
            self.wfile.write(payload)
            return
        self.send_response(405)
        self.end_headers()

    def log_message(self, fmt, *args):
        sys.stdout.write("%s - - [%s] %s\n" % (self.address_string(), self.log_date_time_string(), fmt % args))
        sys.stdout.flush()

    def _write_error(self, status, message):
        payload = json.dumps({"ok": False, "error": message}).encode()
        self.send_response(status)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(payload)))
        self.end_headers()
        self.wfile.write(payload)


def main():
    parser = argparse.ArgumentParser(description="Minimal JSON-RPC relay for Sepolia.")
    parser.add_argument("--listen-host", default="127.0.0.1")
    parser.add_argument("--listen-port", type=int, default=18545)
    parser.add_argument("--upstream", required=True)
    parser.add_argument("--timeout", type=int, default=30)
    args = parser.parse_args()

    RelayHandler.upstream = args.upstream
    RelayHandler.timeout = args.timeout

    server = ThreadingHTTPServer((args.listen_host, args.listen_port), RelayHandler)
    print(
        f"relay listening on http://{args.listen_host}:{args.listen_port} -> {args.upstream}",
        flush=True,
    )
    server.serve_forever()


if __name__ == "__main__":
    main()

File: scripts/sepolia_rpc_relay.js
const http = require("node:http");
const https = require("node:https");
const { URL } = require("node:url");

const upstream = process.env.UPSTREAM_RPC_URL;
const listenHost = process.env.RELAY_LISTEN_HOST || "127.0.0.1";
const listenPort = Number(process.env.RELAY_LISTEN_PORT || "18545");
const requestTimeoutMs = Number(process.env.RELAY_TIMEOUT_MS || "30000");

if (!upstream) {
  console.error("UPSTREAM_RPC_URL is required");
  process.exit(1);
}

const upstreamUrl = new URL(upstream);

const server = http.createServer((req, res) => {
  if (req.method === "GET" && req.url === "/healthz") {
    const body = JSON.stringify({ ok: true, upstream });
    res.writeHead(200, {
      "Content-Type": "application/json",
      "Content-Length": Buffer.byteLength(body),
      Connection: "close",
    });
    res.end(body);
    return;
  }

  if (req.method !== "POST") {
    res.writeHead(405, { Connection: "close" });
    res.end();
    return;
  }

  const chunks = [];
  req.on("data", (chunk) => chunks.push(chunk));
  req.on("end", () => {
    const body = Buffer.concat(chunks);
    const proxyReq = https.request(
      {
        protocol: upstreamUrl.protocol,
        hostname: upstreamUrl.hostname,
        port: upstreamUrl.port || 443,
        path: `${upstreamUrl.pathname}${upstreamUrl.search}`,
        method: "POST",
        headers: {
          "Content-Type": req.headers["content-type"] || "application/json",
          "Content-Length": body.length,
          Connection: "close",
        },
        timeout: requestTimeoutMs,
      },
      (proxyRes) => {
        res.writeHead(proxyRes.statusCode || 502, {
          "Content-Type": proxyRes.headers["content-type"] || "application/json",
          Connection: "close",
        });
        proxyRes.pipe(res);
      }
    );

    proxyReq.on("timeout", () => {
      proxyReq.destroy(new Error("upstream timeout"));
    });

    proxyReq.on("error", (error) => {
      const payload = JSON.stringify({ ok: false, error: error.message });
      res.writeHead(502, {
        "Content-Type": "application/json",
        "Content-Length": Buffer.byteLength(payload),
        Connection: "close",
      });
      res.end(payload);
    });

    proxyReq.end(body);
  });

  req.on("error", (error) => {
    const payload = JSON.stringify({ ok: false, error: error.message });
    res.writeHead(400, {
      "Content-Type": "application/json",
      "Content-Length": Buffer.byteLength(payload),
      Connection: "close",
    });
    res.end(payload);
  });
});

server.listen(listenPort, listenHost, () => {
  console.log(`relay listening on http://${listenHost}:${listenPort} -> ${upstream}`);
});
